### ISSUE TITLE
Take a deeper look at our use of workqueue in the controller

### ISSUE THREAD

**Issue by hbarkal** [bookingcom/shipper#120]
While writing tests I encounter a problem in the release_controller_test: since there we also listen to events, I had a test failing because of a lot of re-processing of the same release key. 
Ignoring events in the release controller test made said tests pass, as was replacing adding items to release workqueue from `Add` to `AddRateLimited`.
To make sure, I added to application controller test an option to listen to events. When doing so, 2 tests that used to pass, now failed. Replacing `Add` with `AddRateLimited` in the application workqueue made these tests pass again.

We use [workqueue](https://godoc.org/k8s.io/client-go/util/workqueue) in our controllers.
A workqueue will not process a single item multiple times concurrently, and if an item is added multiple times before it can be processed, it will only be processed once. However, it is allowed for an item to be reenqueued while it is being processed. To add an item to a workqueue we can use `Add(item interface{})` 
We use out workqueues as RateLimiters, initialised with a DefaultControllerRateLimiter. To add an item to a RateLimitingInterface interface implementing queue, we can use `AddRateLimited(item interface{})`, adds an item to the workqueue after the rate limiter says it's ok.

**Comment by hbarkal** [comment]
`AddRateLimited` might cause latency. We should use UpdateStatus instead of Update when possible.

**Comment by ksurent** [comment]
Regarding the issue that you described. We should probably look into what that test does. I think that rate limiting (more specifically, it's probably the backoff which is a part of the default rate limiter) will likely mask some issues but ultimately may make it worse. The rule of thumb is: use `AddRateLimited` when you're retrying after an error, use `Add` when you're reacting to a change. For retry-able errors you'll want to do increasing backoff between consecutive attempts on the same key (with a ceiling, of course). But if the change is coming from an edit, you'll want to apply that immediately because it may very well be the edit to break the retry loop (consider this: a user does `kubectl edit`, sees their deployment not progressing, realises they made a mistake and fixes it with another `kubectl edit`). This was addressed in https://github.com/bookingcom/shipper/commit/73877ac46536b01c90fc4281a1f5dba4876beb26. I did a very poor job explaining the change (my apologies!), but essentially we fixed the scenario I described above (note how I replaced a bunch of `AddRateLimited` calls with plain `Add`).

Now regarding the PR to address the issue. I'm not sure how resyncs got mixed in here in the first place because the original description talks about error retries. The latter has caused real issues and requires attention. But I would like to understand how resyncs can cause rate limiting. Even when we were running with 30s resync period, it seems implausible. Nowadays, I believe our resync period is 5min. But even if we assume that resyncs somehow trigger rate limiting, resync handling in this PR still doesn't make sense. With a couple of exceptions (enqueuing related objects, abusing resyncs for syncing external state) it's always safe to drop resync events because it's guaranteed that the object hasn't changed and that guarantee comes from etcd.
We should've done this long ago but it kinda feel between cracks. It came up in team conversations  but at the time we probably weren't confident enough to make the change (I know I wasn't).
